proposals/gateway-plusplus-phase1.md

```diff
@@ -0,0 +1,122 @@
+- HTTP RPC API
+    - 😡 Is single-tenant. No authentication system.
+        - The RPC API is designed for a single user to manage a remote node, it was not designed and built to be multi-tenant and would look very different if it were.
+    - 😡 Is not always transactional. Many write calls return before they are available.
+
+These interfaces were built with particular user profiles in mind. The users we’re seeing in NFTs (including our own service as a user of IPFS) are not well served by these interfaces and we need a new “project” definition (Gateway++) that is centered around these user needs.
+
+Developers are building applications that have to run in a lot of constrained environments that can’t support a full node. Strengthening the Gateway to serve these needs over HTTP leverages endpoints they are already using and provides the features they need in a stateless protocol that has ubiquitous support.
+
+We are never going to hear “Sorry, I can’t use HTTP in this environment” from a developer.
+
+## Phase 1
+
+We should expect this to be a long term project with additional proposals in due time.
+
+For now, we have a list of high priority tasks that need to be fixed.
+
+In [nft.storage](http://nft.storage) we have the following high priority needs:
+
+- Add the Pinning API to ipfs-cluster.
+- Add transactional CAR file uploads to the Pinning API.
+
+We also need to do the following to kick off the project.
+
+- Document configuring IPFS Gateway to pass Pinning API requests to a cluster or other IPFS node with the Pinning API enabled.
+- Configure and deploy our own gateway with this configuration. (We won’t be handing out auth tokens to anyone who doesn’t already have one, this is just to eat our own dogfood).
+
+#### What does done look like?
+
+Each work item in Phase 1 is something nft.storage needs so we have a builtin "first user" who would have all their requirements satisfied by this proposal.
+
+Once this is deployed you'll be able to write thin client interfaces with only a gateway URL and bearer token as configuration.
+
+####  What does success look like?
+
+Beyond just having satisfied the needs of nft.storage, this should also satisfy the needs that Cloudflare has and we would hopefully see support for IPFS in Cloudflare Workers as a thin
+client to their Gateway.
+
+#### Alternatives
+
+There are alternative approaches to building thin clients. The proposals around changing/improving the RPC API could be designed for this purpose,
+but the RPC interface is a larger API (not very "thin") and wasn't designed for multi-tenant. There would probably be **more** work involved in getting
+the RPC interface to support these users than this proposal.
+
+#### Counterpoints &amp; pre-mortem
+
+While this maps well to where web developers are today, it's not a "pure p2p" approach to solving problems. We're beefing up the ability to rely on large IPFS nodes that end up
+being federated rather than fully decentralized.
+
```

Comment:
The proposed approach results in wasted bandwidth (upload from the user and download from the service provider) when some of the data already exists on the service provider.

This pushes developers away from working with modifiable/appendable data structures which is something we have otherwise been encouraging.

Reply:
See my comment above regarded “wasted bandwidth.”

We don’t get to determine what data structures people use. NFT developers are already trying to use IPFS and we are not meeting their needs. We may wish they had done something different but we’re past the point of being able to determine their pattern of use.